Improve performance

Sae126V opened the pull request:
Resolves GT-689

Sae126V committed:
Initial commit to work around slow SQL queries

This change adds a standalone script to update the local Django database
from the APEL database on a cron. This means that the webpages remain
responsive.

Sae126V committed:
Add pagination to gridsync records

This page returns a large number of results so needs pagination to not
grind to a halt.

tofu-rocketry commented on monitoring/publishing/db_update_sqlite.py:
```diff
@@ -0,0 +1,283 @@
+django.setup()
+
+
+from monitoring.publishing.models import (
```
Imports at the top.

Sae126V replied:
Ahh! This is a special case. Since this file will be picked up by the `cron`. `Cron` is having a hard time finding the imports - It will be NOT able to find the imports.

Thus, We have to first set up the Django environment, and then the cron is happy to pick.

tofu-rocketry replied:
Ok, then using the example from the other comment, can you group the Django import and the setup call in an  `if __name__ == '__main__':` as per https://docs.djangoproject.com/en/2.2/topics/settings/#calling-django-setup-is-required-for-standalone-django-usage

That'll them make it clearer that there's a particular order that's needed.

Sae126V replied:
The example is on to how to use `django.setup()` i.e, We MUST use `danjo.setup()` and NOT from...import

The scenario we are trying is different here.  We can use `django.setup()` in `__name__ == '__main__'` If it is a normal or a simple file that will get executed in the Django environment itself and NOT as a standalone file via CRON.  So we can't do that in our scenario. Again, Cron is having a hard time finding the imports - It is NOT able to find the imports.

tofu-rocketry commented on monitoring/publishing/db_update_sqlite.py:
```diff
@@ -0,0 +1,283 @@
+import sys
+
+import pandas as pd
+import django
```
Think it's only `django.setup` that's used, so can you do a `from...import`?

Sae126V replied:
I don't think we can use that way.

Acc to the [link](https://docs.djangoproject.com/en/2.2/topics/settings/#calling-django-setup-is-required-for-standalone-django-usage).

Calling django.setup() is required for “standalone” Django usage[¶](https://docs.djangoproject.com/en/2.2/topics/settings/#calling-django-setup-is-required-for-standalone-django-usage)

tofu-rocketry replied:
Ok, fair enough, that does seem to be the standard way.

tofu-rocketry commented on monitoring/publishing/db_update_sqlite.py:
```diff
@@ -0,0 +1,283 @@
+import os
+import sys
+
+import pandas as pd
```
Alphabetically `pandas` should be after `django`.

tofu-rocketry commented on monitoring/publishing/db_update_sqlite.py:
```diff
@@ -0,0 +1,283 @@
+    if len(month_string) == 1:
+        month_string = '0' + month_string
+    return year_string + '-' + month_string
+
```
Should have another blank line.

Sae126V replied:
Done

tofu-rocketry commented on monitoring/publishing/db_update_sqlite.py:
```diff
@@ -0,0 +1,283 @@
+            ON b.SiteName = a.SiteName AND b.UpdateTime = a.latest
+            GROUP BY SiteName;
+        """
+        fetchset =  VAnonCloudRecord.objects.using('cloud').raw(sql_query)
```
Watch the spacing around the operator.

Sae126V replied:
Done

tofu-rocketry commented on monitoring/publishing/views.py:
```diff
@@ -207,10 +173,15 @@ def retrieve(self, request, SiteName=None):
 
 
+class GridSiteSyncPagination(PageNumberPagination):
+    page_size = 1000 # Number of items to be fetched per page
```
Two spaces before inline comments, please.

tofu-rocketry commented on monitoring/publishing/db_update_sqlite.py:
```diff
@@ -0,0 +1,283 @@
+    VSyncRecords
+)
+
+summaries_dict_standard = {
```
Are the original versions of these templates in the other file still needed there?

Sae126V replied:
Yes, It is needed in other places. I am using those again as this is a standalone file

tofu-rocketry replied:
Can we import them instead then? I'd rather not maintain the same thing twice.

Sae126V replied:
The reason I added it in the first place is that I wanted to run this file as a standalone file (with no links to the other files, except for the environment) so that the CRON job has a smoother flow. However, I see that from a maintenance perspective, it doesn't make sense. Updated the code.

tofu-rocketry replied:
"Standalone" can mean different things in different places. In the context of a Django app, we just mean that it's not running as part of the Apache server, but it's never going to be run without making use of the existing models from the app and it lives in the same repository anyway.

tofu-rocketry commented on monitoring/publishing/db_update_sqlite.py:
```diff
@@ -0,0 +1,283 @@
+}
+
+
+def fill_summaries_dict(inpDict, row):
```
Could do with some docstrings on these new functions.

Sae126V replied:
This is NOT a new function. Just a clone of original versions. But, I am happy to provide few words, If needed. Updated the function with docstrings as of now.

tofu-rocketry replied:
Thanks for the new docstrings.

tofu-rocketry commented on monitoring/publishing/db_update_sqlite.py:
```diff
@@ -0,0 +1,283 @@
+
+    log.info(
+        "Data retrieval from the database backend(s) is completed and "
+        "successfully synchronized with the local SQLite database."
```
Has it actually been successful if we reach here, in every situation?

Sae126V replied:
Updated the comments with a slightly different context.

tofu-rocketry reviewed (changes requested):
Mostly style comments for now. Will have a deeper look later on.

Sae126V committed:
Improve styling and comments, and reuse functions

tofu-rocketry commented on monitoring/publishing/db_update_sqlite.py:
```diff
@@ -0,0 +1,302 @@
+import pandas as pd
+
+
+BASE_DIR = os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
+
+# Find the root and the Django project
+sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), '..', '..')))
```
Aren't these two paths equivalent? Wouldn't it be neater to do something like this?

```suggestion
BASE_DIR = os.path.abspath(os.path.join(os.path.dirname(__file__), '..', '..'))

sys.path.append(BASE_DIR)
```

Sae126V replied:
Sure. Done

tofu-rocketry commented on monitoring/publishing/db_update_sqlite.py:
```diff
@@ -0,0 +1,222 @@
+    summaries_dict_standard,
+    syncrecords_dict_standard,
+    correct_dict,
+    determine_sync_status,
```
I think this function is no longer used in `monitoring.publishing.views`, so can you move it into this script?

Sae126V replied:
Done in 5e94025b9fdb39add36eff68c9813d60810c8056

tofu-rocketry commented on monitoring/publishing/db_update_sqlite.py:
```diff
@@ -0,0 +1,222 @@
+# Find the root and the Django project
+sys.path.append(BASE_DIR)
+
+# Set up Django settings to run this `db_update_sqlite.py` as standalone file
+os.environ.setdefault("DJANGO_SETTINGS_MODULE", "monitoring.settings")
+
+# Initialize and setup Django
+django.setup()
```
Can you move the the `import django` in just before the `django.setup()` and then encapsulate all these lines,  in an `if __name__ == "__main__":`  please?

```suggestion
if __name__ == "__main__":
    # Set up to run this `db_update_sqlite.py` script as a standalone file
    os.environ.setdefault("DJANGO_SETTINGS_MODULE", "monitoring.settings")
    import django
    django.setup()
```

Sae126V replied:
We CAN'T do that here, Adrian. The script won't run. 

We always get Traceback (most recent call last):
  File "monitoring/publishing/db_update_sqlite.py", line 24, in <module>
    from monitoring.publishing.models import (

django.core.exceptions.ImproperlyConfigured: Requested setting INSTALLED_APPS, but settings are not configured. You must either define the environment variable DJANGO_SETTINGS_MODULE or call settings.configure() before accessing settings

tofu-rocketry replied:
Ok, is that just an ordering issue then? Could the import got before the `os.environ` or is there another line that depends on this?

tofu-rocketry replied:
Also, as I said on Slack, if the` if...main` bit wouldn't t work, can you at least group the lines I've suggested together (just without the `if` encapsulation)? Then at least it's clearer what we're doing.

Sae126V replied:
I have removed  if encapsulation. But, Still I can't alter the order of the it.

tofu-rocketry replied:
Sorry, I meant the _suggested_ encapsulation, not the `if __name__ == "__main__":` at the bottom of the file around line 213. I'll revert that bit of the change and get this merged.

tofu-rocketry replied:
I've put the if __name__ back in at the end of the file.

Sae126V committed:
Add log separators and move function to script

The function is no-longer used in views, so move it to the script it's
used in.

tofu-rocketry reviewed (approved):
Good job. Sorry for going round in circles on the import and Django setup parts.
